gax/src/main/java/com/google/api/gax/rpc/ApiClientHeaderProvider.java

```diff
@@ -29,188 +29,175 @@
  */
 package com.google.api.gax.rpc;
 
-import com.google.api.client.util.Lists;
 import com.google.api.core.BetaApi;
 import com.google.api.gax.core.GaxProperties;
-import com.google.common.base.Joiner;
-import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableMap;
-import java.util.ArrayList;
-import java.util.List;
+import java.io.Serializable;
 import java.util.Map;
 
 /**
  * Implementation of HeaderProvider that provides headers describing the API client library making
  * API calls.
  */
 @BetaApi("The surface for customizing headers is not stable yet and may change in the future.")
-public class ApiClientHeaderProvider implements HeaderProvider {
-  private static final String DEFAULT_VERSION = "";
-
-  private final String apiClientHeaderLineKey;
-  private final String clientLibName;
-  private final String clientLibVersion;
-  private final String generatorName;
-  private final String generatorVersion;
-  private final List<String> apiClientHeaderLineData;
-  private final String googleCloudResourcePrefix;
+public class ApiClientHeaderProvider implements HeaderProvider, Serializable {
+  private static final long serialVersionUID = -8876627296793342119L;
+
   private final Map<String, String> headers;
 
-  private ApiClientHeaderProvider(Builder builder) {
-    this.apiClientHeaderLineKey = builder.apiClientHeaderLineKey;
-    this.clientLibName = builder.clientLibName;
-    this.clientLibVersion = builder.clientLibVersion;
-    this.generatorName = builder.generatorName;
-    this.generatorVersion = builder.generatorVersion;
-    this.apiClientHeaderLineData = builder.apiClientHeaderLineData;
-    this.googleCloudResourcePrefix = builder.googleCloudResourcePrefix;
-    this.headers = generateHeaders();
+  protected ApiClientHeaderProvider(Builder builder) {
+    ImmutableMap.Builder<String, String> headersBuilder = ImmutableMap.builder();
+
+    if (builder.getApiClientHeaderKey() != null) {
+      StringBuilder apiClientHeaderValue = new StringBuilder();
+      // Order of tokens matters!!!
+      appendToken(apiClientHeaderValue, builder.getJvmToken());
+      appendToken(apiClientHeaderValue, builder.getClientLibToken());
+      appendToken(apiClientHeaderValue, builder.getGeneratedLibToken());
+      appendToken(apiClientHeaderValue, builder.getGeneratedRuntimeToken());
+      appendToken(apiClientHeaderValue, builder.getTransportToken());
+      if (apiClientHeaderValue.length() > 0) {
+        headersBuilder.put(builder.getApiClientHeaderKey(), apiClientHeaderValue.toString());
+      }
+    }
+
+    if (builder.getResourceHeaderKey() != null && builder.getResourceToken() != null) {
+      headersBuilder.put(builder.getResourceHeaderKey(), builder.getResourceToken());
```

Comment:
It is confusing that "token" here applies exclusively to the value, not the whole token.

Reply:
There are 3 components of a header(is true for any header):
1) Header key (or name, `google-cloud-resource-prefix` in this example)
2) Header token. Which, optionally (not true for this example) can have 2 more components:
    2.1) Token name (for example `gapic`, for the header with key (1) `x-goog-api-client`
    2.2) Token version (the semantic version)
In this particular case, the token is not the "versioned one", so it sets token directly (without splitting it on name and version).